### PR TITLE
Feature/refactor data flow analysis small fixes

### PR DESCRIPTION
**Pull request by kindlich**
A bunch of minor changes ^^

### CODE REVIEW

File: Parser/src/main/java/org/openzen/zenscript/parser/expression/ParsedExpressionVariable.java [hunk 1]
```diff
@@ -40,7 +40,7 @@ public CompilingExpression compile(ExpressionCompiler compiler) {
 
 	@Override
 	public CompilingExpression compileKey(ExpressionCompiler compiler) {
-		if (!typeArguments.isEmpty())
+		if (typeArguments != null && !typeArguments.isEmpty())
```
Comment (kindlich) [inline Parser/src/main/java/org/openzen/zenscript/parser/expression/ParsedExpressionVariable.java:43]: 💭 Question:
Is there a semantic difference between `null` and empty type arguments?
Currently, the parser parses `<>` as empty and null if no `<` is encountered

File: CodeModel/src/main/java/org/openzen/zenscript/codemodel/type/builtin/MapTypeSymbol.java [hunk 2]
```diff
@@ -87,8 +88,16 @@ public ResolvedType resolve(TypeID[] typeArguments) {
...
+		members.iterator(new IteratorInstance(
+				type,
+				new TypeID[]{type.keyType},
+				mapper.map(type, BuiltinMethodSymbol.ITERATOR_ASSOC_KEYS)
+		));
+		members.iterator(new IteratorInstance(
+				type,
+				new TypeID[]{type.keyType, type.valueType},
+				mapper.map(type, BuiltinMethodSymbol.ITERATOR_ASSOC_KEY_VALUES)
+		));
```
Comment (kindlich) [inline CodeModel/src/main/java/org/openzen/zenscript/codemodel/type/builtin/MapTypeSymbol.java:100]: 💭 Discussion:
One one hand I believe this is the proper fix here.

On the other hand, we might need to adapt `MemberSet$Builder#method`:
Currently, that method either puts the member into the target's staticMethods or instanceMethods.
However, for primitives we just add all BuiltinMethodSymbol s as `.method()`s, so maybe something could be registered here as a type that does not make sense?

File: JavaBytecodeCompiler/src/main/java/org/openzen/zenscript/javabytecode/compiler/JavaForeachWriter.java [hunk 3]
```diff
@@ -90,6 +90,8 @@ private void handleArray(final int z, final JavaLocalVariableInfo arrayTypeInfo)
 		javaWriter.dup();
 		javaWriter.loadInt(z);
 
+		// ToDo: atm, statement.list.type can also be Optional[ArrayTypeId]
+		//   Does this need to be handled here or should the call already be wrapped in an unwrapOptional expression?
```
Comment (kindlich) [inline JavaBytecodeCompiler/src/main/java/org/openzen/zenscript/javabytecode/compiler/JavaForeachWriter.java:94]: What the ToDo says: Do we need an if optional check, or should this already be done at an earlier stage of the compiling process?

File: Parser/src/main/java/org/openzen/zenscript/parser/expression/ParsedExpressionVariable.java (hunk 1 above)
Comment (stanhebben) [reply Parser/src/main/java/org/openzen/zenscript/parser/expression/ParsedExpressionVariable.java:43]: There is no semantic difference

File: JavaBytecodeCompiler/src/main/java/org/openzen/zenscript/javabytecode/compiler/JavaForeachWriter.java (hunk 3 above)
Comment (stanhebben) [reply JavaBytecodeCompiler/src/main/java/org/openzen/zenscript/javabytecode/compiler/JavaForeachWriter.java:94]: I'd think that if it's an optional, we have to check for it. We'd just need to decide if it should skip the iteration, or throw an exception.

File: ScriptingEngineTester/src/main/resources/zencode-tests/null-coalesce/null-call-2.zc [hunk 4]
```diff
@@ -2,7 +2,7 @@
 #output: hello
 
 function test(): string? => null;
-function test(): string? => "hELLo";
+function test2(): string? => "hello";
 
 println(test()?.toLowerCase());
-println(test()?.toLowerCase());
+println(test2()?.toLowerCase());
```
Comment (stanhebben) [inline ScriptingEngineTester/src/main/resources/zencode-tests/null-coalesce/null-call-2.zc:8]: Is there a reason to change this test?

Comment (kindlich) [reply ScriptingEngineTester/src/main/resources/zencode-tests/null-coalesce/null-call-2.zc:8]: Well, I doubt the test intended 2 functions with the name `test` to be present, did it?

Comment (stanhebben) [reply ScriptingEngineTester/src/main/resources/zencode-tests/null-coalesce/null-call-2.zc:8]: I understand the name change, but I wanted to double-check that toLowerCase was also properly called, hence the uppercase letters in the test ;)

Comment (kindlich) [reply ScriptingEngineTester/src/main/resources/zencode-tests/null-coalesce/null-call-2.zc:8]: Reverted 👍🏽

**Review by stanhebben — APPROVED** [review]

**Pull request merged**